flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolImpl.java

```diff
@@ -759,37 +759,43 @@ private void checkIdleSlot() {
 		final FlinkException cause = new FlinkException("Releasing idle slot.");
 
 		for (AllocatedSlot expiredSlot : expiredSlots) {
-			final AllocationID allocationID = expiredSlot.getAllocationId();
-			if (availableSlots.tryRemove(allocationID) != null) {
-
-				log.info("Releasing idle slot [{}].", allocationID);
-				final CompletableFuture<Acknowledge> freeSlotFuture = expiredSlot.getTaskManagerGateway().freeSlot(
-					allocationID,
-					cause,
-					rpcTimeout);
-
-				FutureUtils.whenCompleteAsyncIfNotDone(
-					freeSlotFuture,
-					componentMainThreadExecutor,
-					(Acknowledge ignored, Throwable throwable) -> {
-						if (throwable != null) {
-							if (registeredTaskManagers.contains(expiredSlot.getTaskManagerId())) {
-								log.debug("Releasing slot [{}] of registered TaskExecutor {} failed. " +
-										"Trying to fulfill a different slot request.", allocationID, expiredSlot.getTaskManagerId(),
-									throwable);
-								tryFulfillSlotRequestOrMakeAvailable(expiredSlot);
-							} else {
-								log.debug("Releasing slot [{}] failed and owning TaskExecutor {} is no " +
-									"longer registered. Discarding slot.", allocationID, expiredSlot.getTaskManagerId());
-							}
-						}
-					});
+
+			if (availableSlots.tryRemove(expiredSlot.getAllocationId()) != null) {
+				releaseSlotToTaskManager(expiredSlot, cause);
 			}
 		}
 
 		scheduleRunAsync(this::checkIdleSlot, idleSlotTimeout);
 	}
 
+	private void releaseSlotToTaskManager(AllocatedSlot expiredSlot, FlinkException cause) {
+		final AllocationID allocationID = expiredSlot.getAllocationId();
+		log.info("Releasing idle slot [{}].", allocationID);
+
+		final CompletableFuture<Acknowledge> freeSlotFuture = expiredSlot.getTaskManagerGateway().freeSlot(
+				allocationID,
+				cause,
+				rpcTimeout);
+
+		FutureUtils.whenCompleteAsyncIfNotDone(
+				freeSlotFuture,
+				componentMainThreadExecutor,
+				(Acknowledge ignored, Throwable throwable) -> {
+					if (throwable != null) {
+						if (throwable instanceof TimeoutException) {
+							log.debug("Releasing slot [{}] of registered TaskExecutor {} timeout. " +
+											"Trying to release it again.",
+									allocationID, expiredSlot.getTaskManagerId(), throwable);
+							releaseSlotToTaskManager(expiredSlot, cause);
```

Comment:
I think it would be better to not unconditionally repeat until success and instead stop after `n` retries. We could use `FutureUtils#retryWithDelay` for that purpose which already contains the required logic. The problem is that this could go on forever in case that the TM failed terminally.

Reply:
Hi @tillrohrmann , thanks for reviewing it. Only adding a slot report between JM and TM seems can not fix the problems. It only solve the part how JM notice TM that a slot it will not use any more if the freeSlot request timeout. But it may happen that TM has released the slot but the request in Jm timeout. So JM may assign it to another execution, but RM may assign it to someone else. The conflict between JM and Tm can not be solved by slot report. 

I think one solution is that we could enhance the check in my fix, when the freeSlot request fails, if the taskManager not exist, it will drop the slot. else if TimeoutException, it will retry. if SlotNotFoundException, it will drop the slot. if other exception such as ConnectClosed, it will tryFulfillSlotRequestOrMakeAvailable. 

Adding TaskSubmissionException together with slot report may could also solve the problem. I think this is question which behavior we choose when a request timeout in distributed system. One principal is that we retry until get a clear answer, another principal could be that we sync state in a background thread to reach non-conflict.

Reply:
I think you are not right @shuai-xu. Based on the heartbeat from the JM to the TM, the TM could free all slots which the TM thinks are still assigned to the JM and fail those which the JM thinks it still owns but does not (via `JobMasterGateway#failSlot`). Alternatively, the TM could send its own slot report back to the JM as the heartbeat's payload so that both sides can reconcile.

I think in general it is not a good idea to retry things forever in distributed systems because you never know how long an outage may last. It could also be the case that only the `freeSlot` messages get lost for whatever reason. Instead the far better approach is to retry it for some time/attempts and then have another means to converge to a consistent view over the system state (e.g. by receiving a special exception saying that one does not own the slot anymore or having a periodic heartbeat which allows you to reconcile the state).

Reply:
@tillrohrmann , Personally, I prefer retrying to periodically heartbeat for status conflict in distributed system. As firstly adding more and more informations to heartbeat will heavy the load of master, especially when it has thousands of task managers. And most report are not necessary as the conflict will not always happen. And secondly, it makes the duration time of conflict depends on the interval of heartbeat, it we extend the interval of heartbeat to 1 minute for example, it means job master may need 1 minute to reconcile wi task manager.
However, if you insist on using heartbeat report, I will change it according to your suggestion.

Reply:
I think we should add the slot report from the JM to the TM as the means to reconcile diverging state. I would suggest to do the following:

1. Simply drop slot and don't reuse slot if `freeSlot` fails. This should cover most cases correctly.
2. Resolve lost `freeSlot` call by sending slot report from the JM to the TM (this should happen rarely)
2.1 TM will free all slots it thinks are still being used by the JM but are not according to the slot report
2.2 TM will fail all slots on the JM which the JM thinks it still owns but does not

It is important to account for in flight offered slots when receiving a slot report from a JM because the slot report might not include them. It could be enough to simply check the active slots and match them with the report. 

Since the rare case (loss of `freeSlot`) is solved with the heartbeat the heartbeat interval should not be a super big problem. Only if the heartbeat interval is longer than the slot request timeout, this could lead to problems because then the newly requested slot might not be fulfillable (in the standalone case where one has only a limited number of slots). WDYT?

Reply:
All right, I will change it as your suggestion.